### PR TITLE
Move dev15.5.x to latest IOperation analyzers for dogfooding

### PR DESCRIPTION
**Pull request by mavasani**
The primary purpose of this PR is to enable dogfooding of the latest compiler toolset, IOperation APIs and analyzers based on these APIs. The PR is for dev15.5.x rather than master because once this change is merged, Roslyn.sln developers will need one of the latest VS builds with Roslyn bits no older than from the last week for analyzers to execute fine in the IDE live analysis. This is only possible for internal developers, and master still needs to support previous VS builds for community work.

### CODE REVIEW

**Review by jasonmalinowski — APPROVED** [review]

**Comment by jasonmalinowski** [comment]
How is this being reconciled with #22380 which implied we _can't_ do this?

**Comment by jasonmalinowski** [comment]
One observation though: since this is in the dev15.5.x branch...@jaredpar what branch is being used for the source build stuff? Is it master or something else? Because having two branches might alleviate the conflict.

**Comment by mavasani** [comment]
If #22380 is a strict requirement, I think one option might be to target this change to 15.6 - I feel it should be okay for us to dogfood the new APIs on a slightly less active branch, but we do want some level of dogfooding in Jenkins. Also tagging @jinujoseph

**Comment by mavasani** [comment]
One other alternative is to make no changes or restrictions to our toolset compiler in the Roslyn repo, but instead have a separate Jenkins job that always uses a shipped toolset compiler to build the latest sources. @jaredpar will this work?

**Comment by mavasani** [comment]
I sync'ed with @jinujoseph and for now we have decided to rebase the PR to post 15.5 so we can get some dogfooding going.

**Review by jinujoseph — APPROVED** [review]

**Comment by mavasani** [comment]
@jasonmalinowski @dotnet/roslyn-infrastructure This PR is targeting post-dev15.5-contrib now. However, I see a `WIP — Waiting for status to be reported` job which doesn't seem to be triggered at all. How do I get past it or force execute it?

**Comment by jcouv** [comment]
>         <LangVersion>7.1</LangVersion>

Would you mind upgrading the language to 7.2, while you're here? Thanks

---
Refers to: build/Targets/Settings.props:238 in 75687b5. [](commit_id = 75687b52623f5873991ea5a161cbafb22c81fa82, deletion_comment = False)

**Comment by mavasani** [comment]
> Would you mind upgrading the language to 7.2, while you're here? Thanks

Done.

File: src/Compilers/Core/Portable/CodeAnalysisResources.Designer.cs [hunk 1]
```diff
@@ -20,7 +20,7 @@ namespace Microsoft.CodeAnalysis {
     // class via a tool like ResGen or Visual Studio.
     // To add or remove a member, edit your .ResX file then rerun ResGen
     // with the /str option, or rebuild your VS project.
-    [global::System.CodeDom.Compiler.GeneratedCodeAttribute("System.Resources.Tools.StronglyTypedResourceBuilder", "4.0.0.0")]
+    [global::System.CodeDom.Compiler.GeneratedCodeAttribute("System.Resources.Tools.StronglyTypedResourceBuilder", "15.0.0.0")]
```
Comment (jcouv) [inline src/Compilers/Core/Portable/CodeAnalysisResources.Designer.cs]: I see this flip-flopping across multiple PRs. Does someone know where this comes from and what is the correct value moving forward?

File: build/Targets/Settings.props [hunk 2]
```diff
@@ -17,12 +17,12 @@
...
     <!-- Disable analyzers if NuGet deviated from the expected restore location (e.g. by changing case on a case-sensitive file system) -->
     <UseRoslynAnalyzers Condition="!Exists('$(NuGetPackageRoot)\Microsoft.CodeQuality.Analyzers\$(RoslynDiagnosticsNugetPackageVersion)\analyzers\dotnet\cs\Analyzer.Utilities.dll')">False</UseRoslynAnalyzers>
-    <RoslynDiagnosticsPropsFilePath>$(NuGetPackageRoot)/microsoft.net.roslyndiagnostics/$(RoslynDiagnosticsNugetPackageVersion)/roslyn/Microsoft.Net.RoslynDiagnostics.props</RoslynDiagnosticsPropsFilePath>
+    <RoslynDiagnosticsPropsFilePath>$(NuGetPackageRoot)/microsoft.net.roslyndiagnostics/$(RoslynDiagnosticsNugetPackageVersion)/build/Microsoft.Net.RoslynDiagnostics.props</RoslynDiagnosticsPropsFilePath>
     <RoslynPortableTargetFrameworks>net461;netcoreapp2.0</RoslynPortableTargetFrameworks>
     <RoslynPortableTargetFrameworks46>net46;netcoreapp2.0</RoslynPortableTargetFrameworks46>
     <RoslynPortableRuntimeIdentifiers>win7;win7-x64</RoslynPortableRuntimeIdentifiers>
  
-    <Features>strict,IOperation</Features>
+    <Features>strict</Features>
```
Comment (jcouv) [inline build/Targets/Settings.props]: I looks like the logic to handle features (in `IsFeatureEnabled`) is being removed by this PR. Either it should stay or this last feature value can be removed.

**Comment by jcouv** [comment]
>         private ImmutableDictionary<string, string> _features;

Are there any uses of `_features` left?

---
Refers to: src/Compilers/CSharp/Portable/CSharpParseOptions.cs:22 in f5fc55f. [](commit_id = f5fc55f808fdfcfd2ff04d2fccf3b9506fb54a98, deletion_comment = False)

File: src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb [hunk 3]
```diff
@@ -2677,11 +2677,6 @@ baddate:
         End Function
 
         Private Shared Function CheckFeatureAvailability(parseOptions As VisualBasicParseOptions, feature As Feature) As Boolean
-            Dim featureFlag = feature.GetFeatureFlag()
-            If featureFlag IsNot Nothing Then
-                Return parseOptions.Features.ContainsKey(featureFlag)
```
Comment (jcouv) [inline src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb]: Same question on `Features` here.

**Review by jcouv — COMMENTED** [review]
Done with review pass on compiler changes (iteration 4).

File: src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb (hunk 3 above)
Comment (jcouv) [reply src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb]: Tagging @gafter who mentioned we should not remove too much features machinery, even if no features at the moment.

Comment (mavasani) [reply src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb]: This was actually removed a while back in dev15.5.x as it was unused code once we removed IOperation feature flag. I can revert the changes that removed the machinery with this PR.

Comment (mavasani) [reply src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb]: Note this PR is just cherry picking existing commits that have gone in dev15.5 branch: https://github.com/dotnet/roslyn/pull/22684/commits/27dc60f5f82fff906914b7770b57ca4deaacce5c and https://github.com/dotnet/roslyn/pull/22684/commits/75687b52623f5873991ea5a161cbafb22c81fa82

File: src/Compilers/CSharp/Portable/CSharpParseOptions.cs [hunk 4]
```diff
@@ -202,11 +202,6 @@ internal override void ValidateOptions(ArrayBuilder<Diagnostic> builder)
 
         internal bool IsFeatureEnabled(MessageID feature)
         {
-            string featureFlag = feature.RequiredFeature();
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/CSharpParseOptions.cs]: Please do not remove this from code from the compiler. It is intended to facilitate experimental code residing in branches of the compiler, even though the product code does not contain experiments.

File: src/Compilers/CSharp/Portable/CSharpResources.resx [hunk 5]
```diff
@@ -4401,9 +4401,6 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
   <data name="ERR_FeatureNotAvailableInVersion7" xml:space="preserve">
     <value>Feature '{0}' is not available in C# 7.0. Please use language version {1} or greater.</value>
   </data>
-  <data name="ERR_FeatureIsExperimental" xml:space="preserve">
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/CSharpResources.resx]: Please do not remove this from code from the compiler. It is intended to facilitate experimental code residing in branches of the compiler, even though the product code does not contain experiments.

File: src/Compilers/CSharp/Portable/Errors/ErrorCode.cs [hunk 6]
```diff
@@ -1294,7 +1294,7 @@ internal enum ErrorCode
         ERR_EncodinglessSyntaxTree = 8055,
         // ERR_AccessorListAndExpressionBody = 8056, Deprecated in favor of ERR_BlockBodyAndExpressionBody
         ERR_BlockBodyAndExpressionBody = 8057,
-        ERR_FeatureIsExperimental = 8058,
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Errors/ErrorCode.cs]: Please do not remove this from code from the compiler. It is intended to facilitate experimental code residing in branches of the compiler, even though the product code does not contain experiments.

File: src/Compilers/CSharp/Portable/Errors/MessageID.cs [hunk 7]
```diff
@@ -178,23 +178,6 @@ public static LocalizableErrorArgument Localize(this MessageID id)
             return new LocalizableErrorArgument(id);
         }
 
-        // Returns the string to be used in the /features flag switch to enable the MessageID feature.
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Errors/MessageID.cs]: Please do not remove this from code from the compiler. It is intended to facilitate experimental code residing in branches of the compiler, even though the product code does not contain experiments.

File: src/Compilers/CSharp/Portable/Parser/Lexer.cs [hunk 8]
```diff
@@ -901,17 +901,6 @@ private void CheckFeatureAvailability(MessageID feature)
                 return;
             }
 
-            string requiredFeature = feature.RequiredFeature();
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Parser/Lexer.cs]: Please do not remove this from code from the compiler. It is intended to facilitate experimental code residing in branches of the compiler, even though the product code does not contain experiments.

File: src/Compilers/CSharp/Portable/Parser/SyntaxParser.cs [hunk 9]
```diff
@@ -1058,30 +1058,16 @@ protected TNode CheckFeatureAvailability<TNode>(TNode node, MessageID feature, b
             }
 
             var featureName = feature.Localize();
-            var requiredFeature = feature.RequiredFeature();
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Parser/SyntaxParser.cs]: Please do not remove this from code from the compiler. It is intended to facilitate experimental code residing in branches of the compiler, even though the product code does not contain experiments.

File: src/Compilers/Test/Utilities/CSharp/TestOptions.cs [hunk 10]
```diff
@@ -83,34 +81,6 @@ public static CSharpParseOptions WithReplaceFeature(this CSharpParseOptions opti
         {
             return options;
         }
-
-        internal static CSharpParseOptions WithExperimental(this CSharpParseOptions options, params MessageID[] features)
```
Comment (gafter) [inline src/Compilers/Test/Utilities/CSharp/TestOptions.cs]: Please do not remove this from code from the compiler. It is intended to facilitate experimental code residing in branches of the compiler, even though the product code does not contain experiments.

File: src/Compilers/Test/Utilities/VisualBasic/TestOptions.vb [hunk 11]
```diff
@@ -31,28 +29,4 @@ Friend Module TestOptionExtensions
     Public Function WithStrictFeature(options As VisualBasicParseOptions) As VisualBasicParseOptions
         Return options.WithFeatures(options.Features.Concat(New KeyValuePair(Of String, String)() {New KeyValuePair(Of String, String)("Strict", "true")}))
     End Function
-
-    <Extension()>
-    Friend Function WithExperimental(options As VisualBasicParseOptions, ParamArray features As Feature()) As VisualBasicParseOptions
```
Comment (gafter) [inline src/Compilers/Test/Utilities/VisualBasic/TestOptions.vb]: Please do not remove this from code from the compiler. It is intended to facilitate experimental code residing in branches of the compiler, even though the product code does not contain experiments.

File: src/Compilers/VisualBasic/Portable/Parser/ParserFeature.vb [hunk 12]
```diff
@@ -32,24 +32,13 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
...
         Tuples
-        IOperation
         InferredTupleNames
         LeadingDigitSeparator
         NonTrailingNamedArguments
         PrivateProtected
     End Enum
 
     Friend Module FeatureExtensions
-        <Extension>
```
Comment (gafter) [inline src/Compilers/VisualBasic/Portable/Parser/ParserFeature.vb]: Please do not remove this from code from the compiler. It is intended to facilitate experimental code residing in branches of the compiler, even though the product code does not contain experiments.

File: src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb (hunk 3 above)
Comment (gafter) [reply src/Compilers/VisualBasic/Portable/Scanner/Scanner.vb]: Please do not remove this from code from the compiler. It is intended to facilitate experimental code residing in branches of the compiler, even though the product code does not contain experiments.

**Comment by mavasani** [comment]
@gafter @jcouv I have reverted the commits that removed experimental framework. Please review again.

**Comment by mavasani** [comment]
@jcouv Ubuntu builds are failing with the below error:

```
error CS1617: Invalid option '7.2' for /langversion; must be ISO-1, ISO-2, Default, Latest or a valid version in range 1 to 7.1. [/mnt/j/workspace/dotnet_roslyn/post-dev15.5-contrib/ubuntu_14_debug_prtest/src/Compilers/Core/Portable/CodeAnalysis.csproj]
```

**Comment by jcouv** [comment]
@mavasani Ok. Take the 7.2 back. I'll investigate when I'm back in office. I suspect the problem is that we use a different package reference for .Net Core.

**Comment by jcouv** [comment]
@mavasani Anything holding up merging this?

**Comment by jcouv** [comment]
Ping @mavasani Once this is merged, I'll enable C# 7.2.

**Pull request merged**